README
$Header$

move files atomicly.

Compile:
	make

Usage:
	mvatom [options] oldname newname
	mvatom [options] -d directory name...
	mvatom [options] -b name
	dirlist
	dirlist DIR

About:
	See DESCRIPTION

-Tino
webmaster@scylla-charybdis.com
http://www.scylla-charybdis.com/tools.php

$Log$
Revision 1.2  2008-05-07 15:12:17  tino
dirlist added

Revision 1.1  2006-07-22 23:47:58  tino
First version for testing



ChangeLog
This ChangeLog is retired.  See individual Git commits.

2016-11-30  Valentin Hilbig  <tino@medusa.03.softkill.org>

	* cmpanddel.sh: use "IFS=''" for "read -r" to allow spaces at the
	end of filenames, too.
	(cmpsoftlink): LF at the end of softlinks previously were ignored,
	now 'a' -> 'b' and 'a' -> 'b\n' are considered to be different.

2008-05-28  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* mvatom.c (do_rename_away, main): Option -c

2008-05-26  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* cmpanddel.sh (unpf): Output bugfix, src/dst interchanged

2008-05-20  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* dirlist.c (main, dirlist): options -0 and -

2008-05-13  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* mvatom.c (do_mvaway): bugfix for option -s and minor resouce
	leakage fixed
	(do_rename_backup, main): m_append (option -a) added

2008-04-26  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* mvatom.c (do_rename_backup): If destination exists output now
	makes more sense.

2007-10-01  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* cmpanddel.sh (unpf): made more readable

2006-10-04    <tino@firebird.03.softkill.net>

	* mvatom.c: tino_va_* routines

2006-09-27    <tino@firebird.03.softkill.net>

	* mvatom.c (do_mvaway): now reads lines
	(do_mkdirs): added for verbose output

2006-08-12    <tino@firebird.03.softkill.net>

	* mvatom.c (main): option -o added
	(read_dest, main): option -l prepared, option -0 disabled

2006-08-01    <tino@firebird.03.softkill.net>

	* mvatom.c (do_rename_backup, main): option -p added


mvatom.c
/*
 * Atomic file move utility.
 *
 * Copyright (C)2006-2011 Valentin Hilbig <webmaster@scylla-charybdis.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA.
 */

#define TINO_NEED_OLD_ERR_FN
#include "tino/filetool.h"
#include "tino/ex.h"
#include "tino/getopt.h"
#include "tino/buf_line.h"

#include "mvatom_version.h"

static int		errflag;
static int		m_backup, m_ignore, m_nulls, m_lines, m_relative, m_quiet, m_verbose, m_mkdirs, m_append;
static const char	*m_dest, *m_source, *m_backupdir;
#if 0
static int		m_unsafe, m_force;
static const char	*m_tmpdir;
#endif


/**********************************************************************/

static void
verror_fn(const char *prefix, TINO_VA_LIST list, int err)
{
  if (!m_quiet)
    tino_verror_ext(list, err, "mvatom %s", prefix);
  if (!m_ignore)
    exit(1);
  errflag	= 1;
}

static void
verbose(const char *s, ...)
{
  tino_va_list	list;

  if (!m_verbose)
    return;

  tino_va_start(list, s);
  vprintf(s, tino_va_get(list));
  tino_va_end(list);
  printf("\n");
}


/**********************************************************************/

/* This actually is a hack.
 *
 * We only have one single operation active at a time.  So we can use
 * a static buffer here which keeps the intermediate string.
 */
static const char *
get_src(const char *name)
{
  static TINO_BUF	buf;

  if (!m_source)
    return name;

  tino_buf_resetO(&buf);
  tino_buf_add_sO(&buf, m_source);
  tino_buf_add_sO(&buf, name);
  return tino_buf_get_sN(&buf);
}

static void
do_mkdirs(const char *path, const char *file)
{
  switch (tino_file_mkdirs_forfileE(path, file))
    {
    case -1:
      tino_err("failed: mkdir for %s%s%s", path ? path : "", path ? "/" : "", file);
      break;

    case 1:
      verbose("mkdir for %s%s%s", path ? path : "", path ? "/" : "", file);
      break;
    }
}

static const char *
read_dest(void)
{
  static TINO_BUF buf;

  if (!m_nulls && !m_lines)
    {
      tino_err("missing option -l or -0 to read stdin");
      return 0;
    }
  return tino_buf_line_readE(&buf, 0, (m_nulls ? 0 : '\n'));
}

static int
rename_noclobber(const char *name, const char *to)
{
  /* We need a rename which is
   * a) atomic
   * b) fails if *to exists
   *
   * There is renameat2() which can be used.  But is that portable enough?
   * Anyway, use it for now.
   *
   * Note that rename() has a very good purpose:
   *
   * It renames the source into the destination such, that the destination never is gone missing.
   * So it is even double atomic.
   *
   * We should support that, too, but the major idea is not to harm data,
   * so atomically and unconditionally replacing a destination is not what we want.
   *
   * To really be able to atomically replace a destination without doubt, following is needed:
   * a) create a reference to the destination by hardlinking it to the backup location
   * b) atomically replace the destination, if it still resembles the hardlinked file
   *
   * So what we need is a rename() call with 3 arguments:
   * Source	Path to replace the Dest
   * Dest	Path to be replaced
   * Ref	Handle which must be identical to Dest, else the rename fails
   *
   * Also we need a way to "materialize a link to a file handle", so the way would be:
   *
   * - open dest for read(!), giving handle Dest
   * - hardlink Dest (the handle!) to the backup location, so materialize Dest to the filesystem
   * - rename src to dest with Dest as safeguard (this here is missing)
   * - close Dest
   *
   * Why open for read must suffice?  Due to SeLinux.
   * It may be that I am allowed to hardlink, rename and read the file, but writing is forbiden.
   *
   * Well, this also shows another missing link:  Opening a file for reference only, so neither read nor write.
   *
   * Note for Linux:  You probably can hardlink /proc/fd/HANDLE to the given destination,
   * so materializing a file descriptor is presend there (even that this is weird).
   */
  return renameat2(AT_FDCWD, name, AT_FDCWD, to, RENAME_NOREPLACE);
}

static int
do_rename(const char *name, const char *to)
{
  if (rename_noclobber(name, to))
    {
      tino_err("cannot rename %s -> %s", name, to);
      return 1;
    }
  verbose("rename: %s -> %s", name, to);
  return 0;
}

/* rename away *name, that is
 * move *name to a *rename
 * possibly into directory (option -c)
 * possibly with an backup extension appended (option -b or option -a)
 */
static int
do_rename_away(const char *name, const char *rename)
{
  char	*tmp1, *tmp2, ret;

  tmp1	= 0;
  if (m_backupdir)	/* option -c present	*/
    rename	= tmp1	= tino_file_glue_pathOi(NULL, 0, m_backupdir, tino_file_filenameptr_constO(rename));
  tmp2	= 0;
  if (!tino_file_notexistsE(rename))
    {
      if (!m_append && !m_backup)
        {
          tino_err("existing backup destination: %s", rename);
          tino_freeO(tmp1);
          return 1;
        }
      rename	= tmp2	= tino_file_backupnameNi(NULL, 0, rename);
    }
  else if (m_mkdirs)
    do_mkdirs(NULL, rename);
  ret = do_rename(name, rename);
  tino_freeO(tmp1);
  tino_freeO(tmp2);
  return ret;
}

static int
do_rename_backup(const char *old, const char *new)
{
  const char	*src;

  src	= get_src(old);

  /* Try to move, skips a lot of syscalls in the most common situation
   */
  if (!rename_noclobber(src, new))
    {
      verbose("rename: %s -> %s", src, new);
      return 0;
    }

  /* Try to figure out what happened	*/

  if (tino_file_notexistsE(src))
    {
      tino_err("missing old name for rename: %s", src);
      return 1;
    }
  errno	= 0;	/* Do not report errors in case there is no error	*/

  if (!tino_file_notexistsE(new))
    {
      /* destination exists	*/
      if (m_append)
        {
          /* just rename away the source (option -a) to the given destination	*/
          return do_rename_away(src, new);
        }
      if (!m_backup && !m_backupdir)
        {
          /* destination exists and we cannot backup the destination	*/
          tino_err("existing destination: %s", new);
          return 1;
        }
      /* rename away the destination	*/
      if (do_rename_away(new, new))
        return 1;
    }
  else if (m_mkdirs)
    do_mkdirs(NULL, new);
  return do_rename(src, new);
}


/**********************************************************************/

static int
do_mvaway(const char *name)
{
  const char	*src;

  src	= get_src(name);
  if (tino_file_notexistsE(src))
    {
      tino_err("missing file to move away: %s", src);
      return 1;
    }
  return do_rename_away(src, src);
}

static int
mvaway(const char *name)
{
  int	ret = 0;

  if (strcmp(name, "-"))
    ret	= do_mvaway(name);
  else
    while ((name=read_dest())!=0)
      ret |= do_mvaway(name);
  return ret;
}


/**********************************************************************/

/* leaks memory	*/
static const char *
do_relative(const char *old, const char *dest)
{
  if (!m_relative)
    return dest;
  return tino_file_glue_pathOi(NULL, 0, tino_file_dirnameOi(NULL, 0, old), tino_file_skip_root_constN(dest));
}

/* leaks memory	*/
static int
mvrename(const char *old, const char *new)
{
  return do_rename_backup(old, do_relative(old, new));
}


/**********************************************************************/

static int
do_mvdest(const char *name)
{
  const char	*targ;
  char		*dest;
  int		ret;

  if (m_relative)
    {
      targ	= tino_file_skip_root_constN(name);
      do_mkdirs(m_dest, targ);
    }
  else
    targ	= tino_file_filenameptr_constO(name);
  dest	= tino_file_glue_pathOi(NULL, 0, m_dest, targ);
  ret	= do_rename_backup(name, dest);
  tino_freeO(dest);
  return ret;
}

static int
mvdest(const char *name)
{
  int	ret = 0;
  if (m_mkdirs<2 && tino_file_notdirE(m_dest))
    {
      tino_err((tino_file_notexistsE(m_dest) ? "missing destination directory: %s" : "existing destination not a directory: %s"), m_dest);
      return 1;
    }
  if (strcmp(name, "-"))
    ret	= do_mvdest(name);
  else
    while ((name=read_dest())!=0)
      ret	|= do_mvdest(name);
  return ret;
}


/**********************************************************************/

static int
is_directory_target(const char *name)
{
  const char	*tmp;

  tmp	= tino_file_filenameptr_constO(name);
  if (*tmp && strcmp(tmp, ".") && strcmp(tmp, ".."))
    return 0;
  return !tino_file_notdirE(name);
}

int
main(int argc, char **argv)
{
  int	argn, m_original;

  tino_verror_fn	= verror_fn;

  argn	= tino_getopt(argc, argv, 1, 0,
                      TINO_GETOPT_VERSION(MVATOM_VERSION)
                      " name...\n"
                      "	if name is - lines are read from stdin.\n"
                      "	rename:		%s -r oldname newname\n"
                      "	move:		%s -d dir name...\n"
                      "	rename away:	%s -ab name\n"
                      "	convenience:	alias mv='mvatom -o'",

                      TINO_GETOPT_USAGE
                      "h	this help"
                      ,

                      TINO_GETOPT_FLAG
                      "0	(This option is 'number zero', not a big O!)\n"
                      "		read 0 terminated lines from stdin\n"
                      "		example: find . -type f -print0 | mvatom -0ab -"
                      , &m_nulls,

                      TINO_GETOPT_FLAG
                      "a	Append .~#~ to source if destination exists\n"
                      "		This has less race conditions than option -b.\n"
                      "		Use with option -b to 'move away' a file."
                      , &m_append,

                      TINO_GETOPT_FLAG
                      "b	Backup existing destination to .~#~\n"
                      "		Cannot be used with option -a except for 'rename away' feature.\n"
                      "		On errors this might leave you with a renamed destination!"
                      , &m_backup,

                      TINO_GETOPT_STRING
                      "c dir	Create backups in the given directory.\n"
                      "		This moves an existing destination into the given dir,\n"
                      "		possibly renaming it according to options -a and -b"
                      , &m_backupdir,

                      TINO_GETOPT_STRING
                      "d dir	target (Destination) directory to move files into"
                      , &m_dest,
#if 0
                      TINO_GETOPT_FLAG
                      "f	Force overwrite of destination, needs unsafe mode (option -u)\n"
                      "		This directly calls rename() per move and therefor atomically\n"
                      "		replaces (overwrites) existing destinations unconditionally.\n"
                      "		Also this needs only 1 syscall per move.\n"
                      , &m_force,
#endif
                      TINO_GETOPT_FLAG
                      "i	Ignore (common) errors"
                      , &m_ignore,

                      TINO_GETOPT_FLAG
                      "l	read Lines from stdin, enables '-' as argument\n"
                      "		example: find . -print | mvatom -lb -"
                      , &m_lines,

                      TINO_GETOPT_FLAG
                      "o	Original behavior if directory is the last target\n"
                      "		The last argument must end in a / or must be . or .."
                      , &m_original,

                      TINO_GETOPT_FLAG
                      TINO_GETOPT_MAX
                      "p	create missing Parent directories (for -r)\n"
                      "		Give twice to create for -d, too"
                      , &m_mkdirs,
                      2,

                      TINO_GETOPT_FLAG
                      "q	silently fail"
                      , &m_quiet,

                      TINO_GETOPT_FLAG
                      "r	append Relative source path to dest (needs no -p)\n"
                      "		To rename within a directory without moving use:\n"
                      "		mvatom -r /path/to/file/a b"
                      , &m_relative,

                      TINO_GETOPT_STRING
                      "s src	append the given Src prefix, for an usage like:\n"
                      "		( cd whatever; ls -1; ) | mvatom -l -s whatever/ -d todir -\n"
                      "		The source prefix is not a directory, it's a literal prefix"
                      , &m_source,
#if 0
                      TINO_GETOPT_STRING
                      "t dir	use the given Temporary working directory for safe mode\n"
                      "		The directory must be in the same filesystem as source and dest.\n"
                      "		Without this option such a directory is created.  In unsafe mode\n"
                      "		this directory is not used, but it still must be present."
                      , &m_tmpdir,

                      TINO_GETOPT_FLAG
                      "u	Unsafe mode, needs 2 syscalls instead of 3 by using rename()\n"
                      "		This mode was previously the default.  It first checks the\n"
                      "		destination for existence and then uses rename() to move a file.\n"
                      "		If the destination is created between those two syscalls, this\n"
                      "		cannot be detected and therefor rename() overwrites (destroys) it."
                      , &m_unsafe,
#endif
                      TINO_GETOPT_FLAG
                      "v	verbose"
                      , &m_verbose,

                      NULL);

  if (argn<=0)
    return 1;

  if (m_original && !m_dest && argn+1<argc && is_directory_target(argv[argc-1]))
    m_dest	= argv[--argc];
  if (!m_dest && m_backup && m_append && argc==argn+1)
    {
      mvaway(argv[argn]);
      return errflag;
    }
  if (m_backup && m_append)
    tino_err("Options -a and -b cannot be used together this way");
  if (m_dest)
    {
      while (argn<argc)
        mvdest(argv[argn++]);
    }
  else if (argc==argn+2)
    mvrename(argv[argn], argv[argn+1]);
  else
    tino_err("Option -d not present: Rename needs 2 arguments; For 'move away' now use options -ab");
  return errflag;
}


cmpanddel.sh
#!/bin/bash
#
# Compare directory with other directories and delete files which match.
#
# This Works is placed under the terms of the Copyright Less License,
# see file COPYRIGHT.CLL.  USE AT OWN RISK, ABSOLUTELY NO WARRANTY.

set -e

STDOUT() { local e=$?; printf '%q' "$1"; printf ' %q' "${@:2}"; printf '\n'; return $e; }
STDERR() { STDOUT "$@" >&2; }
OOPS() { STDERR OOPS: "$@"; exit 23; }
INTERNAL() { OOPS internal error: "$@"; }
x() { "$@"; }
o() { x "$@" || OOPS fail $?: "$@"; }

maketmpdir()
{
if [ -n "$1" ]
then
	tmpdir="$1"
	[ ! -d "$1" ] && o mkdir "$1"
else
	tmpdir="$(mktemp -d tmpcmp.XXXXXX)"
fi

pf="$tmpdir/file"
pl="$tmpdir/location"
trap '[ -e "$pf" ] || rm -f "$pl"; rmdir "$tmpdir"' 0
}

usage()
{
[ 2 -le $# ] && return
cat <<EOF
Usage: "$(basename "$0")" directory-to-cleanup directory-to-compare..
	Temporary directory may be given in env CMPANDDEL_TMP
	You can give multiple compare-directories to check for files,
	the first one with the given target wins.
EOF
exit 42
}

warn()
{
el="$(tput el || :)$(tput cr || :)"
SRC="$1"
DSTS=("${@:2}")
cat <<EOF

This script is for lazy experienced sysadmins and comes with mvatom:
http://www.scylla-charybdis.com/tool.php?tool=mvatom

This script removes files from the first directory which match files
in the second directory.  It checks softlinks, too.  Be sure that the
directory to delete is quiescent, this is, nobody else modifies files
or changes anything.  Also never run two of this scripts in parallel
on the same directory, you have been warned!

If this script fails have a look in the tempdir created, there are two
entries, "file" and "location".  Location tells you where file came
from.  Move it back manually.  (No warranty.  Use at own risk.  etc.)

Deletes  from "$SRC"
EOF

for x in "${DSTS[@]}"
do
	echo "Compares from \"$x\""
done
echo
echo -n "Press return to continue: "
read
}

# Try to find the given file in the DSTS
: get_dst file
get_dst()
{
local a
dst=/does-not-exist
for a in "${DSTS[@]}"
do
	dst="$a/$1"
	[ -e "$dst" ] && return
done
:
}

# Protect a file by moving it into the TMPDIR
# The idea behind this is, that if you compare source with destination,
# the source vanishes, such that the destination vanishes, too.
# This effectively protects against accidents like: cmp x x && rm x
prot()
{
[ ! -e "$pf" ] || INTERNAL "$pf" exists

echo -n "$1" > "$pl"
o mvatom "$1" "$pf"
unpf="$1"
}

# Move file back from compare-directory to where it was
unp()
{
[ -n "$unpf" ] || INTERNAL variable not set
[ ! -e "$unpf" ] || INTERNAL "$unpf" exists
[ ".$unpf." = ".$(cat "$pl"; echo .)" ] || INTERNAL "$unpf" does not match "$pl"

# Do not bail out on error in case something vanished.
# In that case it might leave debris behind for manual cleanup.
# This is better than to give up early.  (Gives up in prot())
x mvatom "$pf" "$unpf"
unpf=""
}

# Print progress output
show()
{
x="$(( ${#2}<70 ? 0 : ${#2}-70 ))"
printf '%q %q%s' "$1" "${2:$x}" "$el"
}

getlink()
{
find "$1" -type d -prune -o -type l -printf '%l'
echo x	# prevent POSIX bug removing arbitrary number of \n from output
}

cmpsoftlink()
{
show link "$1"
get_dst "$1"

if [ ! -L "$dst" ]
then
	if [ -e "$dst" ]
	then
		echo "$dst is no soflink!"
	else
		echo "$dst does not exist"
	fi
	return
fi

prot "$SRC/$1"

b="$(getlink "$pf")"
c="$(getlink "$dst")"
if [ ".$b" != ".$c" ]
then
	unp
	echo "$dst: softlinks mismatch: $SRC/$1"
	return
fi

rm -f "$pf"
unpf=""
}

cmpfile()
{
show file "$1"
get_dst "$1"

if [ ! -e "$dst" ]
then
	echo "$dst does not exist"
	return
fi

prot "$SRC/$1"

if [ -L "$dst" -o ! -f "$dst" ]
then
	unp
	echo "$dst is no normal file!"
	return
fi

if ! cmp "$pf" "$dst"
then
	unp
	echo "$SRC/$1 mismatch."
	return
fi

rm -f "$pf"
unpf=""
}

cmpsock()
{
show sock "$1"
get_dst "$1"

if [ ! -S "$dst" ]
then
	if [ -e "$dst" ]
	then
		echo "$dst is no socket!"
	else
		echo "$dst does not exist"
	fi
	return
fi

prot "$SRC/$1"

if [ ! -S "$dst" ]
then
	unp
	echo "$dst socket vanished!"
	return
fi

rm -f "$pf"
unpf=""
}

cmpfifo()
{
show fifo "$1"
get_dst "$1"

if [ ! -p "$dst" ]
then
	if [ -e "$dst" ]
	then
		echo "$dst is no fifo!"
	else
		echo "$dst does not exist"
	fi
	return
fi

prot "$SRC/$1"

if [ ! -p "$dst" ]
then
	unp
	echo "$dst fifo vanished!"
	return
fi

rm -f "$pf"
unpf=""
}

getmajorminor()
{
stat -c %t,%T "$1"
}

cmpspecial()
{
show "$1" "$4"
get_dst "$4"

if [ ! -$2 "$dst" ]
then
	if [ -e "$dst" ]
	then
		echo "$dst is no $3 special!"
	else
		echo "$dst does not exist"
	fi
	return
fi

prot "$SRC/$4"

if [ ! -$2 "$dst" ]
then
	unp
	echo "$dst $3 special vanished!"
	return
fi

b="$(getmajorminor "$pf")"
c="$(getmajorminor "$dst")"
if [ ".$b" != ".$c" ]
then
	unp
	echo "$dst: $3 special mismatch ('$c' vs. '$b'): $SRC/$4"
	return
fi

rm -f "$pf"
unpf=""
}

cmpblock()
{
cmpspecial "blk " b block "$1"
}

cmpchar()
{
cmpspecial char c char "$1"
}

cmpdir()
{
show "dir " "$1"
get_dst "$1"

if [ ! -e "$dst" ]
then
	echo "$dst does not exist"
	return
fi

prot "$SRC/$1"

if [ -L "$dst" -o ! -d "$dst" ]
then
	unp
	echo "$dst is no directory!"
	return
fi

if ! rmdir "$pf"
then
	unp
	echo "$pf not empty"
	return
fi
unpf=""
}

usage "$@"
warn "$@"
maketmpdir "$CMPANDDEL_TMP"

find "$SRC" -type f -printf '%P\0' -o -type d -name "$tmpdir" -prune |
while IFS='' read -rd '' a
do
	cmpfile "$a"
done

find "$SRC" -type l -printf '%P\0' -o -type d -name "$tmpdir" -prune |
while IFS='' read -rd '' a
do
	cmpsoftlink "$a"
done

find "$SRC" -type s -printf '%P\0' -o -type d -name "$tmpdir" -prune |
while IFS='' read -rd '' a
do
	cmpsock "$a"
done

find "$SRC" -type p -printf '%P\0' -o -type d -name "$tmpdir" -prune |
while IFS='' read -rd '' a
do
	cmpfifo "$a"
done

find "$SRC" -type b -printf '%P\0' -o -type d -name "$tmpdir" -prune |
while IFS='' read -rd '' a
do
	cmpblock "$a"
done

find "$SRC" -type c -printf '%P\0' -o -type d -name "$tmpdir" -prune |
while IFS='' read -rd '' a
do
	cmpchar "$a"
done

find "$SRC" -depth -type d -name "$tmpdir" -prune -o -type d -printf '%P\0' |
while IFS='' read -rd '' a
do
	[ -z "$a" ] || cmpdir "$a"
done



dirlist.c
/* $Header$
*
* Primitive directory lister
*
* Copyright (C)2008-2009 Valentin Hilbig <webmaster@scylla-charybdis.com>
*
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License as
* published by the Free Software Foundation; either version 2 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful, but
* WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
* General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
* 02110-1301 USA.
*
* $Log$
* Revision 1.4  2009-04-09 18:05:57  tino
* dist
*
* Revision 1.3  2008-10-16 19:42:50  tino
* Multiargs and options -a -d -l -m -u -o
*
* Revision 1.2  2008-05-21 17:56:53  tino
* Options
*
* Revision 1.1  2008-05-07 15:12:17  tino
* dirlist added
*/

#include "tino/main_getini.h"
#include "tino/err.h"
#include "tino/filetool.h"
#include "tino/slist.h"

#include <dirent.h>

#include "mvatom_version.h"

static int f_nulls, f_buffered, f_parent, f_nodot, f_source, f_one, f_debug, f_read, f_readz, f_recurse;
static unsigned f_set, f_unset, f_set_any, f_unset_any;
static TINO_SLIST	subdirs;

static void
putmode(unsigned mode)
{
  if (mode)
    putmode(mode/8);
  putchar('0'+(mode&7));
}

static long
get_mode(const char *dir, const char *file)
{
  static const char	*lastname;
  static unsigned	lastmode;
  const char		*tmp;
  tino_file_stat_t	st;
  long			ret;

  tmp	= tino_file_glue_pathOi(NULL, (size_t)0, dir, file);
  if (lastname && !strcmp(tmp, lastname))
    ret	= lastmode;
  else if (tino_file_lstatE(tmp, &st))
    {
      TINO_ERR1("ETTDI102B %s: cannot stat", tmp);
      ret	= -1;
    }
  else
    ret	= st.st_mode;

  if (lastname)
    tino_free_constO(lastname);
  lastname	= tmp;
  return ret;
}

static int
do_stat(const char *dir, const char *subdir, const char *name)
{
  long	mode;
  char	*tmp;

  tmp	= tino_file_glue_pathOi(NULL, (size_t)0, subdir, name);
  mode	= get_mode(dir, tmp);
  if (mode<0)
    {
      tino_freeO(tmp);
      return -1;
    }
  if (f_recurse && S_ISDIR(mode) && ( !name || name[0]!='.' || ((name[1] && name[1]!='.') || name[2]) ))
    tino_glist_add_ptr(subdirs, tmp);
  else
    tino_freeO(tmp);

  if (f_set && ((unsigned)mode&f_set)!=f_set)
    return 1;
  if (f_unset && ((unsigned)mode&f_unset)!=0)	/* same as: if (st.st_mode&f_unset) return 1;	*/
    return 1;
  if (f_set_any && ((unsigned)mode&f_set_any)==0)
    return 1;
  if (f_unset_any && ((unsigned)mode&f_unset_any)==f_unset_any)
    return 1;

  if (f_debug)
    {
      putmode((unsigned)mode);
      putchar(' ');
    }
  return 0;
}

static void
out(const char *name)
{
  fputs(name, stdout);
  putchar(f_nulls ? 0 : '\n');
  if (!f_buffered)
    fflush(stdout);
}

static int
do_dirlist(const char *dir, const char *subdir, const char *both, int stats)
{
  DIR           *dp;
  struct dirent *d;

  if ((dp=opendir(both))==0)
    {
      TINO_ERR1("ETTDI100B %s: cannot open", both);
      return 0;
    }
  while (errno=0, (d=readdir(dp))!=0)
    {
      if (d->d_name[0]=='.')
	{
	  if (d->d_name[1]==0 || (d->d_name[1]=='.' && d->d_name[2]==0))
	    {
	      if (!f_parent)
		continue;
	    }
	  else if (f_nodot)
	    continue;
	}

      if ((f_recurse || stats) && do_stat(dir, subdir, d->d_name))
	continue;

      if (f_source || subdir)
	{
	  const char	*tmp;

	  tmp	= tino_file_glue_pathOi(NULL, (size_t)0, (f_source ? both : subdir), d->d_name);
	  out(tmp);
	  tino_free_constO(tmp);
	}
      else
	out(d->d_name);
    }
  if (errno)
    {
      TINO_ERR1("ETTDI103B %s: read error", both);
      closedir(dp);
      return 0;
    }
  if (closedir(dp))
    {
      TINO_ERR1("ETTDI101B %s: close error", both);
      return 0;
    }
  return 1;
}

static int
do_dirlist_step(const char *dir, const char *sub, int stats)
{
  const char	*tmp;
  int		ret;

  tmp	= tino_file_glue_pathOi(NULL, (size_t)0, dir, sub);
  ret	= do_dirlist(dir, sub, tmp, stats);
  tino_free_constO(tmp);
  return ret;
}

static int
do_dirlist_start(const char *dir, int stats)
{
  const char	*sub;

  if (!do_dirlist_step(dir, NULL, stats))
    return 0;

  while ((sub=tino_slist_get(subdirs))!=0)
    {
      do_dirlist_step(dir, sub, stats);
      tino_free_constO(sub);
    }
  return 1;	/* Directory found, even if it contains errors in subdirectories	*/
}

static void
do_dirlist1(const char *dir, int stats)
{
  if (do_dirlist_start(dir, stats) && f_one)
    exit(0);
}

static void
dirlist(const char *dir, void *user)
{
  subdirs	= tino_slist_new();

  if (f_buffered)
    setvbuf(stdout, NULL, _IOFBF, BUFSIZ);
  if (!dir)
    dir=".";
  if (!f_debug)
    do_dirlist1(dir, (f_set || f_unset || f_set_any || f_unset_any));
  else if (!do_stat(dir, NULL, NULL))
    {
      out(dir);
      if (f_recurse)
	do_dirlist_start(dir, (f_set || f_unset || f_set_any || f_unset_any));
    }
  tino_slist_destroy(subdirs);
  subdirs	= 0;
}

int
main(int argc, char **argv)
{
  return tino_main_if(dirlist,
		      NULL,	/* user	*/
		      NULL,
		      argc, argv,
		      0, -1,
		      TINO_GETOPT_VERSION(MVATOM_VERSION)
		      " [directory..]\n"
		      "\tA primitive directory lister, lists all names in a dir,\n"
		      "\tskips . and .. by default."
		      ,

		      TINO_GETOPT_FLAG
		      "0	write NUL terminated lines (instead of LF)"
		      , &f_nulls,

		      TINO_GETOPT_UNSIGNED
		      "a mode	any given bit set in Mode. (see -m)"
		      , &f_set_any,

		      TINO_GETOPT_FLAG
		      "d	debug mode, print octal mode in front of filename\n"
		      "		Does no directlry list, needs a stat() call"
		      , &f_debug,

		      TINO_GETOPT_FLAG
		      "f	full buffered IO (no flushs after each line)"
		      , &f_buffered,
#if 0
		      TINO_GETOPT_FLAG
		      "i	on file arguments, read targets from it, '-' for stdIn"
		      , &f_read,
#endif
		      TINO_GETOPT_UNSIGNED
		      "l mode	given bits must be unset in Mode. (see -m)"
		      , &f_unset,

		      TINO_GETOPT_UNSIGNED
		      "m mode	given bits must be set in Mode. (see also -l -a -u)\n"
		      "		Prefix with 0 for octal, needs additional stat() calls.\n"
		      "		If -a -l -m -u is used together, all must apply."
		      , &f_set,

		      TINO_GETOPT_FLAG
		      "n	no dot-files, hides files starting with a ."
		      , &f_nodot,

		      TINO_GETOPT_FLAG
		      "o	one arg mode, this lists the first available directory"
		      , &f_one,

		      TINO_GETOPT_FLAG
		      "p	list . and .."
		      , &f_parent,

		      TINO_GETOPT_FLAG
		      "r	recurse subdirectories"
		      , &f_recurse,

		      TINO_GETOPT_FLAG
		      "s	add source path to output"
		      , &f_source,

		      TINO_GETOPT_UNSIGNED
		      "u mode	any given bit unset in Mode. (see -m)"
		      , &f_unset_any,
#if 0
		      TINO_GETOPT_FLAG
		      "z	on -i read NUL terminated input (else lines)"
		      , &f_readz,
#endif
		      NULL);
}
